python_scripts/02_numerical_pipeline_cross_validation.py

```diff
@@ -0,0 +1,151 @@
+
+model = make_pipeline(StandardScaler(), LogisticRegression())
+
+# %% [markdown]
+# ## The need for cross-validation
+#
+# In the previous notebook, we split the original data into a training set and a
+# testing set. But the score of a model will in general depend on the way we make
+# such a split. One downside of doing a single split is that it does not give
+# any information about the variability of the score. Another downside is that, in
+# a setting where the amount of data is small, the data available for training
+# and testing will be even smaller after splitting.
+#
+# Instead, we can use cross-validation. Cross-validation consists of repeating
+# the procedure such that the training and testing sets are different each
+# time. Generalization performance metrics are collected for each repetition and
+# then aggregated. As a result we can get an estimate of the variability of the
+# model's generalization performance.
+#
+# Note that there exists several cross-validation strategies, each of them
+# defines how to repeat the `fit`/`score` procedure. In this section, we will
+# use the K-fold strategy: the entire dataset is split into `K` partitions. The
+# `fit`/`score` procedure is repeated `K` times where at each iteration `K - 1`
+# partitions are used to fit the model and `1` partition is used to score. The
+# figure below illustrates this K-fold strategy.
+#
+# ![Cross-validation diagram](../figures/cross_validation_diagram.png)
+#
+# ```{note}
+# This figure shows the particular case of **K-fold** cross-validation strategy.
+# For each cross-validation split, the procedure trains a model on all the red
+# samples and evaluate the score of the model on the blue samples.
+# As mentioned earlier, there is a variety of different cross-validation
+# strategies. Some of these aspects will be covered in more detail in future
+# notebooks.
+# ```
+#
+# For each cross-validation split, the procedure trains a model on all the red
+# samples and evaluate the score of the model on the blue samples.
+# Cross-validation is therefore computationally intensive because it requires
+# training several models instead of one.
+#
+# In scikit-learn, the function `cross_validate` allows to do cross-validation
+# and you need to pass it the model, the data, and the target. Since there
+# exists several cross-validation strategies, `cross_validate` takes a
+# parameter `cv` which defines the splitting strategy.
+
+# %%
+# %%time
+from sklearn.model_selection import cross_validate
+
+model = make_pipeline(StandardScaler(), LogisticRegression())
+cv_result = cross_validate(model, data_numeric, target, cv=5)
+cv_result
+
+# %% [markdown]
+# The output of `cross_validate` is a Python dictionary, which by default
+# contains three entries: (i) the time to train the model on the training data
+# for each fold, (ii) the time to predict with the model on the testing data
+# for each fold, and (iii) the default score on the testing data for each fold.
```

Comment:
Let's use an item list to make this more visual and move the most important things first:

```suggestion
# contains three entries:
# - (i) the score of each model evaluated on the testing data for each fold,
# - (ii) the time to train each model on the training data for each fold,
# - (iii) the time to predict with each model on the testing data for each fold.
```

also, let's not talk about "the" model when during CV as we typically fit 5 or 10 models.